posix/posix.c

```diff
@@ -1068,68 +1071,68 @@ int posix_fstat(int fd, struct stat *buf)
 {
 	TRACE("fstat(%d)", fd);
 
-	msg_t msg;
 	open_file_t *f;
+	msg_t msg;
 	int err;
 
-	if (!(err = posix_getOpenFile(fd, &f))) {
-		hal_memset(buf, 0, sizeof(struct stat));
-		hal_memset(&msg, 0, sizeof(msg_t));
+	if ((err = posix_getOpenFile(fd, &f)) < 0)
+		return err;
 
-		buf->st_dev = f->ln.port;
-		buf->st_ino = (int)f->ln.id; /* FIXME */
-		buf->st_rdev = f->oid.port;
+	hal_memset(buf, 0, sizeof(struct stat));
+	hal_memset(&msg, 0, sizeof(msg_t));
 
-		if (f->type == ftRegular) {
-			msg.type = mtGetAttr;
-			hal_memcpy(&msg.i.attr.oid, &f->oid, sizeof(oid_t));
-			msg.i.attr.val = 0;
-
-			do {
-				msg.i.attr.type = atMTime;
-				if (!(err = proc_send(f->oid.port, &msg)))
-					buf->st_mtime = msg.o.attr.val;
-				else break;
-
-				msg.i.attr.type = atATime;
-				if (!(err = proc_send(f->oid.port, &msg)))
-					buf->st_atime = msg.o.attr.val;
-				else break;
-
-				msg.i.attr.type = atCTime;
-				if (!(err = proc_send(f->oid.port, &msg)))
-					buf->st_ctime = msg.o.attr.val;
-				else break;
-
-				msg.i.attr.type = atLinks;
-				if (!(err = proc_send(f->oid.port, &msg)))
-					buf->st_nlink = msg.o.attr.val;
-				else break;
-
-				msg.i.attr.type = atMode;
-				if (!(err = proc_send(f->oid.port, &msg)))
-					buf->st_mode = msg.o.attr.val;
-				else break;
-
-				msg.i.attr.type = atUid;
-				if (!(err = proc_send(f->oid.port, &msg)))
-					buf->st_uid = msg.o.attr.val;
-				else break;
-
-				msg.i.attr.type = atGid;
-				if (!(err = proc_send(f->oid.port, &msg)))
-					buf->st_gid = msg.o.attr.val;
-				else break;
-
-				msg.i.attr.type = atSize;
-				if (!(err = proc_send(f->oid.port, &msg)))
-					buf->st_size = msg.o.attr.val;
-				else break;
-			}
-			while (0);
-		}
-		else {
-			switch (f->type) {
+	buf->st_dev = f->ln.port;
+	buf->st_ino = (int)f->ln.id; /* FIXME */
```

Comment:
Would be better to be more elaborate what exactly needs to be fixed

Reply:
The `FIXME` comment is from some older commit. I guess narrowing conversion here was the issue (`f.ln.id` is of `id_t` type - `u64` on most platforms).